### PR TITLE
Generate triggers

### PR DESCRIPTION
**Pull request by SimonRichardson**
Triggers were written whereby any update to the database would cause a new insert into the change log, even if the value was identical. This causes a lot of churn in the change log. We can generate the triggers using the columns from the actual database schema. The triggers will always be correct and always up to date.

Part of the change set includes fixing watchers and watcher tests:

### Cloud watcher

The watch cloud test wasn't working how it was expected. The upsert was just triggering a new change in the old code, prior to this patch. The cloud-config is currently immutable, I'll open a bug to fix this. As the config is not updated and the trigger prevents a noop change then the test fails. The solution is to then update a field on the cloud that will cause a change. In this case, the endpoint was updated.

Bug about cloud-config: https://bugs.launchpad.net/juju/+bug/2066410

### Model config watcher

The model config watch wasn't correct once the generated triggers were added. Firstly, we deleted all of the keys and then re-added them again even if they were the same. This caused a lot of churn in the changestream. The fix was just to ensure that we inserted, updated, and deleted rows correctly.

The second problem was that we would see the initial changes from bootstrap and then see them again when the change stream was processing requests. This might be ok for some watchers (I highly doubt that), but the concept of watchers are that, watch changes and then react. For most things that take information from the API server, the initial
event is thrown away, so it's pointless trying to get that. It's better to just read the state of the world and then respond to changes.

The model_config watcher will now emit an empty initial event and then subsequent events will tell you what has changed.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ make rebuild-triggers
```

```sh
$ juju bootstrap lxd test
$ juju add-model default
$ juju deploy ubuntu
```

## Links

**Jira card:** [JUJU-6055](https://warthogs.atlassian.net/browse/JUJU-6055)

[JUJU-6055]: https://warthogs.atlassian.net/browse/JUJU-6055?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by manadart — COMMENTED** [review]
The generated triggers are wrong. You want to write on update if this **OR** that changes, not only for both this **AND** that.

**Comment by SimonRichardson** [comment]
@manadart there is a slight wart with this, as it's compiled into non-test code, the lack of the file or even a typo means you can't create the file again without commenting out the code that depends on it.

**Comment by manadart** [comment]
I was thinking about this over the weekend. The inequality checks don't work if either of the values are `NULL`, and we can't use arbitrary values for `IFNULL` on both sides, because we don't know the valid values for every field.

It looks like we will have might have to do the filtering in the specific watchers.

**Comment by SimonRichardson** [comment]
> I was thinking about this over the weekend. The inequality checks don't work if either of the values are NULL, and we can't use arbitrary values for IFNULL on both sides, because we don't know the valid values for every field.

We can check for that in the DB schema and work through that.

**Comment by SimonRichardson** [comment]
For example `PRAGMA table_info(<table-name>); ` will tell you if it's expecting `NULL` or not. Then we can apply the following rules:

 1. If NOT NULL - `a != b`
 2. If NULL - `(IFNULL(a, b) OR a != b)`

**Review by manadart — COMMENTED** [review]
This is a better option than the esoteric trigger functions. We'll need to ensure it's understood and accepted as an alternative.

File: domain/blockdevice/state/state.go [hunk 1]
```diff
@@ -177,24 +177,7 @@ func (st *State) updateBlockDevices(ctx context.Context, tx *sqlair.TX, machineU
 		fsTypeByName[fsType.Name] = fsType.ID
 	}
 
-	insertQuery := `
-INSERT INTO block_device (uuid, machine_uuid, name, label, device_uuid, hardware_id, wwn, bus_address, serial_id, mount_point, size_mib, filesystem_type_id, in_use)
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply domain/blockdevice/state/state.go:181]: I'm going to rewind this, I was trying to test the null change.

File: generate/triggergen/main.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+	"fmt"
+
+	"github.com/juju/juju/core/database/schema"
+)
+
+{{range .Views}}
+func ChangeLogTriggersFor{{title .Name}}(columnName string, namespaceID int) func() schema.Patch {
+	return func() schema.Patch {
+		return schema.MakePatch(fmt.Sprintf(` + "`" + `
+-- {{title .Name}} for insert trigger
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply generate/triggergen/main.go:286]: No reason

File: domain/schema/controller.go [hunk 3]
```diff
@@ -5,9 +5,12 @@ package schema
 
 import (
 	"github.com/juju/juju/core/database/schema"
+	"github.com/juju/juju/domain/schema/controller"
 )
 
-type tableNamespaceID int
+//go:generate go run ./../../generate/triggergen -db=controller -destination=./controller/triggers.gen.go -package=controller -tables=cloud,cloud_credential,external_controller,controller_config,controller_node,model_migration_status,model_migration_minion_sync,upgrade_info,upgrade_info_controller_node,object_store_metadata_path,secret_backend_rotation,model
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply domain/schema/controller.go:11]: 💯 I'll send an email as well, this one is a big one.

File: domain/schema/changelog.go [hunk 4]
```diff
@@ -4,11 +4,11 @@
 package schema
 
 import (
-	"fmt"
-
 	"github.com/juju/juju/core/database/schema"
 )
 
+type tableNamespaceID = int
```
Comment (SimonRichardson) [inline domain/schema/changelog.go:10]: I think we should rethink this at some point. It doesn't provide strong guarantees between schema and change log watchers.

File: domain/schema/testing/schema.go [hunk 5]
```diff
@@ -29,3 +33,24 @@ func (s *SchemaApplier) Apply(c *gc.C, ctx context.Context, runner database.TxnR
 	c.Assert(err, gc.IsNil)
 	c.Check(changeSet.Post, gc.Equals, s.Schema.Len())
 }
+
+// DumpChangeLog dumps the change log to the test log.
+func DumpChangeLog(ctx context.Context, c *gc.C, runner database.TxnRunner) {
```
Comment (SimonRichardson) [inline domain/schema/testing/schema.go:38]: This is VERY useful for debugging the change log.

File: domain/schema/testing/schema.go [hunk 6]
```diff
@@ -29,3 +33,52 @@ func (s *SchemaApplier) Apply(c *gc.C, ctx context.Context, runner database.TxnR
 	c.Assert(err, gc.IsNil)
 	c.Check(changeSet.Post, gc.Equals, s.Schema.Len())
 }
+
+// DumpChangeLogState dumps the change log to the test log.
+func DumpChangeLogState(ctx context.Context, c *gc.C, runner database.TxnRunner) {
```
Comment (manadart) [inline domain/schema/testing/schema.go:38]: This is not used at present, but it's OK to keep as a utility.

I would omit the context argument though, as it's only used in tests, and just supply `context.Background` to the query calls directly.

**Review by manadart — APPROVED** [review]
You've already got a conflict, but everything looks good.

Probably worth an email to crew about working with it.

File: domain/schema/testing/schema.go (hunk 6 above)
Comment (SimonRichardson) [reply domain/schema/testing/schema.go:38]: Agreed. Will fix.

File: generate/triggergen/main.go [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	databaseschema "github.com/juju/juju/core/database/schema"
+	"github.com/juju/juju/domain/schema"
+	"github.com/juju/juju/internal/database/app"
+	"github.com/juju/juju/internal/database/txn"
+)
+
+func main() {
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply generate/triggergen/main.go:29]: Fair.

**Pull request merged**